generatebundlefile/ecr.go

```diff
@@ -2,46 +2,106 @@ package main
 
 import (
 	"context"
+	"encoding/base64"
 	"fmt"
 	"log"
+	"os/exec"
 	"reflect"
 	"regexp"
 	"strings"
 	"time"
 
 	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/ecr"
+	"github.com/aws/aws-sdk-go-v2/service/ecr/types"
+	ecrtypes "github.com/aws/aws-sdk-go-v2/service/ecr/types"
 	"github.com/aws/aws-sdk-go-v2/service/ecrpublic"
-	"github.com/aws/aws-sdk-go-v2/service/ecrpublic/types"
+	ecrpublictypes "github.com/aws/aws-sdk-go-v2/service/ecrpublic/types"
 	"github.com/aws/aws-sdk-go/aws"
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/pkg/errors"
 
 	api "github.com/aws/eks-anywhere-packages/api/v1alpha1"
 )
 
-type ecrClient struct {
+type ecrPublicClient struct {
 	*ecrpublic.Client
+	AuthConfig     string
+	SourceRegistry string
 }
 
-// NewECRClient Creates a new ECR Client Public client
-func NewECRClient() (*ecrClient, error) {
+type ecrClient struct {
+	*ecr.Client
+	AuthConfig string
+}
+
+// NewECRPublicClient Creates a new ECR Client Public client
+func NewECRPublicClient(creds bool) (*ecrPublicClient, error) {
 	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-east-1"))
 	if err != nil {
 		log.Fatal(err)
 	}
-	ecrClient := &ecrClient{Client: ecrpublic.NewFromConfig(cfg)}
+	ecrPublicClient := &ecrPublicClient{Client: ecrpublic.NewFromConfig(cfg)}
+	if err != nil {
+		return nil, err
+	}
+	if creds {
+		sourceAuthConfig, err := ecrPublicClient.GetPublicAuthConfig()
+		if err != nil {
+			return nil, err
+		}
+		ecrPublicClient.AuthConfig = fmt.Sprintf("%s:%s", sourceAuthConfig.Username, sourceAuthConfig.Password)
+		return ecrPublicClient, nil
+	}
+	return ecrPublicClient, nil
+}
+
+// NewECRClient Creates a new ECR Client Public client
+func NewECRClient(creds bool) (*ecrClient, error) {
+	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-west-2"))
+	if err != nil {
+		log.Fatal(err)
+	}
+	ecrClient := &ecrClient{Client: ecr.NewFromConfig(cfg)}
 	if err != nil {
 		return nil, err
 	}
+	if creds {
+		sourceAuthConfig, err := ecrClient.GetAuthConfig()
+		if err != nil {
+			return nil, err
+		}
+		ecrClient.AuthConfig = fmt.Sprintf("%s:%s", sourceAuthConfig.Username, sourceAuthConfig.Password)
+		return ecrClient, nil
+	}
 	return ecrClient, nil
 }
 
 // Describe returns a list of ECR describe results, with Pagination from DescribeImages SDK request
-func (c *ecrClient) Describe(describeInput *ecrpublic.DescribeImagesInput) ([]types.ImageDetail, error) {
-	var images []types.ImageDetail
+func (c *ecrPublicClient) DescribePublic(describeInput *ecrpublic.DescribeImagesInput) ([]ecrpublictypes.ImageDetail, error) {
+	var images []ecrpublictypes.ImageDetail
 	resp, err := c.DescribeImages(context.TODO(), describeInput)
 	if err != nil {
 		return nil, fmt.Errorf("error: Unable to complete DescribeImagesRequest to ECR public. %s", err)
 	}
 	images = append(images, resp.ImageDetails...)
+	if resp.NextToken != nil {
```

Comment:
When you do a describe call it will pass back a list with 50 items, and either a next token or not. We don't have to do recursion to look for the next token here. We just keep re-calling the function until we don't see a next token to make sure we get all objects.

I think we could implement some sort of image recycling out of band, as there would be recursion issues when this list gets too large (Too many images in the repo).